File: core/src/main/kotlin/app/dopedgoal/core/Goal.kt
package app.dopedgoal.core

/**
 * One task on a goal. Its brick is minted here, at creation, and kept forever —
 * completing the task only decides *where* on the wall the brick lands.
 */
data class Task(
    val id: String,
    val goalId: String,
    val name: String,
    val brick: Brick,
    val completedAt: Long? = null,
) {
    val isComplete: Boolean get() = completedAt != null

    init {
        require(name.isNotBlank()) { "a task needs a name" }
        require(brick.taskId == id) { "brick ${brick.taskId} was not minted for task $id" }
    }
}

enum class Category { HEALTH, LEARNING, CAREER, CREATIVE, SOCIAL, PERSONAL }

/**
 * A goal is a wall. Its task count is fixed at creation, so the wall has a
 * finite, visible end from the first brick onward.
 *
 * Tasks may be completed in any order. [placedCount] is what fills the wall
 * bottom-up; [tasks] stays in the order the user wrote them.
 */
data class Goal(
    val id: String,
    val name: String,
    val category: Category,
    val emoji: String,
    val cohesion: Cohesion,
    val style: WallStyle,
    val banner: Banner,
    val tasks: List<Task>,
    val createdAt: Long,
    val completedAt: Long? = null,
) {
    init {
        require(name.isNotBlank()) { "a goal needs a name" }
        require(tasks.isNotEmpty()) { "a wall needs at least one brick" }
        require(tasks.all { it.goalId == id }) { "every task must belong to goal $id" }
        require(tasks.map { it.id }.toSet().size == tasks.size) { "duplicate task id" }
        require(banner.goalId == id) { "banner was minted for ${banner.goalId}, not $id" }
    }

    val total: Int get() = tasks.size

    val placedCount: Int get() = tasks.count { it.isComplete }

    val isAchieved: Boolean get() = placedCount == total

    /** Bricks in the order they were actually placed — the wall's real build order. */
    val placedBricks: List<Brick>
        get() = tasks.filter { it.isComplete }
            .sortedBy { requireNotNull(it.completedAt) }
            .map { it.brick }
}

File: flutter/lib/models/goal.dart
import 'package:flutter/material.dart';

enum GoalCategory { health, learning, career, creative, social, personal }

extension CategoryLabel on GoalCategory {
  String get label => switch (this) {
        GoalCategory.health => 'HEALTH',
        GoalCategory.learning => 'LEARNING',
        GoalCategory.career => 'CAREER',
        GoalCategory.creative => 'CREATIVE',
        GoalCategory.social => 'SOCIAL',
        GoalCategory.personal => 'PERSONAL',
      };

  Color get dotColor => switch (this) {
        GoalCategory.health => const Color(0xFFC96A5C),
        GoalCategory.learning => const Color(0xFF4A7FC4),
        GoalCategory.career => const Color(0xFF8A6BBF),
        GoalCategory.creative => const Color(0xFFC9639A),
        GoalCategory.social => const Color(0xFF5FA8A0),
        GoalCategory.personal => const Color(0xFF6BA368),
      };
}

/// One task on a goal. Its brick color is fixed at creation; completing the
/// task only decides whether the brick is showing on the wall yet.
class Task {
  Task({
    required this.id,
    required this.name,
    this.completedAt,
  });

  final String id;
  final String name;
  final DateTime? completedAt;

  bool get isComplete => completedAt != null;

  Task copyWith({DateTime? completedAt}) =>
      Task(id: id, name: name, completedAt: completedAt ?? this.completedAt);
}

/// A goal is a wall: a fixed list of tasks/bricks, built up as tasks complete.
class Goal {
  Goal({
    required this.id,
    required this.name,
    required this.category,
    required this.emoji,
    required this.cohesion,
    required this.wallSeed,
    required this.tasks,
    required this.createdAt,
  }) : assert(tasks.isNotEmpty, 'a wall needs at least one brick');

  final String id;
  final String name;
  final GoalCategory category;
  final String emoji;
  final double cohesion;
  final String wallSeed;
  final List<Task> tasks;
  final DateTime createdAt;

  int get total => tasks.length;
  int get placedCount => tasks.where((t) => t.isComplete).length;
  bool get isAchieved => placedCount == total;
  double get progress => total == 0 ? 0 : placedCount / total;
}

/// One completed task, as it shows up in History — a task plus the goal it
/// belongs to, so the log doesn't need to look the goal back up.
class HistoryEntry {
  HistoryEntry({
    required this.task,
    required this.goalName,
    required this.goalEmoji,
    required this.category,
    required this.completedAt,
  });

  final Task task;
  final String goalName;
  final String goalEmoji;
  final GoalCategory category;
  final DateTime completedAt;
}
